### ISSUE TITLE
[DS-510] Add support for buttons

### ISSUE THREAD

**Issue by ghost** [utsushi/utsushi#5]
Forgive me if I'm missing something, but it appears that the DS-510 driver does not have button support. I'd like to set them up but `scanimage -A` show no sensors entries. I assume this means that I cannot use them until the driver supports them.

I'm using the recently released imagescan 3.12.0-1epson4debian8 package from Epson's site.

Thanks for your work.

**Comment by paddy-hack** [comment]
You're not missing anything.  The driver is missing something: push button support.

On the device protocol level, support is there, see `driver/esci/grammar-status.hpp` and the associated `.cpp` file.  It is just not exposed as a (read-only) option by the `compound_scanner` class, probably because it is not clear what the various `event()`s mean.  The push button's `is_duplex()` and `media_size()` bits appear to be understood.

The utsushi code base has been synced with the sources from the version you are using.

Info on which buttons cause what events and patches are of course welcome :smiley: but I understand that that is not quite for the faint of heart.  I don't have a DS-510 (or similar) so I cannot investigate this myself.

**Comment by ghost** [comment]
Thanks for the description. I would be all over sending a patch were I any good at C. However, all I can offer are the aforementioned button event codes. I would be happy to supply them if they would be useful and if I were pushed in the right direction for discovering all of the info.

**Comment by paddy-hack** [comment]
Actually, thinking this through a bit, all that would be needed is a read-only option that queries the device for its push button status and provide the event status.  It doesn't really matter what each individual value means as long as you can determine that a button was pushed.  You can tie the event value (an integer in the [0..3] range, where 0 probably means no button pushed) to whatever action you like.

Of course, this assumes there is something that is polling for that read-only option's value and you have defined the actions you want taken for each of the possible values.  I understand that `scanbd` can be used for this but have zero experience with it.

I kind of busy right now helping out with a `sane-backends` release, but will take a look at adding an option when I find some time.

**Comment by ghost** [comment]
Yes, scanbd essentially runs a daemon that polls and then sends a SANE event through localhost upon push (SANE must be configured to only use `net` and scanbd will handle the actual printer interfaces). A good overview can be read here: https://wiki.archlinux.org/index.php/Scanner_Button_Daemon.

Let me know how I, a pleb with decent admin experience and knowledge in bash, can help.

**Comment by paddy-hack** [comment]
Hate to disappoint you but until the driver gets button query support there isn't anything you can do.  Once the support is there though you can help testing.

**Comment by ghost** [comment]
No problem! I hacked together a different solution in the meantime that works quite well. :)

**Comment by paddy-hack** [comment]
Have you just been staring at this ticket for 6 days until you got a reply :wink: 

I you could explain or put up your hack, it might help other people in the mean time as well.

**Comment by ghost** [comment]
> Have you just been staring at this ticket for 6 days until you got a reply :wink: 

Yes, I bought a 55" TV and mounted it on my wall so I could keep this ticket in my field of view at all times.

Since my ds-510 is single-purpose, I bought a ten-dollar button meant for jokes (The "Big Red Button"). A driver can be easily put together: I found three previous implementations in python, ruby, and C. I prefer C so I went with that:

```
/*
,* From http://ddurdle.blogspot.com/2013/12/using-usb-big-red-button-panic-button.html
,*/

#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define LID_CLOSED 21
#define BUTTON_PRESSED 22
#define LID_OPEN 23

int main(int argc, char **argv)
{
  int fd;
  int i, res, desc_size = 0;
  char buf[256];

  /* Use a udev rule to make this device */
  fd = open("/dev/hidraw0", O_RDWR|O_NONBLOCK);

  if (fd < 0) {
    perror("Unable to open device");
    return 1;
  }

  int prior = LID_CLOSED;

  while (1) {
    memset(buf, 0x0, sizeof(buf));
    buf[0] = 0x08;
    buf[7] = 0x02;

    res = write(fd, buf, 8);
    if (res < 0) {
      perror("write");
      exit(1);
    }

    memset(buf, 0x0, sizeof(buf));
    res = read(fd, buf, 8);

    if (res >= 0) {
      if (prior != BUTTON_PRESSED && buf[0] == BUTTON_PRESSED) {
         printf("Scanning Receipt");
         system("SCANSCRIPT-OR-COMMAND");
         fflush(stdout);
      }
      prior = buf[0];
    }
    usleep(20000); /* Sleep for 20ms*/
  }
}
```

I stripped out the unneeded functionality of detecting lid open/close. I just need a button press :)

I just created a systemd service that runs an imagescan script that performs all of the functionality that I need. I run it as the `scan` user that I created: To allow all users to read the button state, run `chmod 666` on /dev/hidraw0.

```
[Unit]
Description=Driver for the Big Red Button

[Service]
ExecStart=/usr/local/bin/big-red-button1
User=scan

[Install]
WantedBy=multi-user.target
```

Alternatively, one could set up a udev rule (I am using two BRBs: one for a specific scan function). The following creates /dev/big_red_button_1 which links to the proper hidraw device:

```
KERNEL=="hidraw*", ATTRS{idVendor}=="1d34", ATTRS{idProduct}=="000d", ATTRS{busnum}=="4", SYMLINK+="big_red_button_1" MODE="0666", GROUP="plugdev"
```

The above expects the BRB to be plugged in to bus number 4: Adjust to your needs.
